Engine/Graphics/OpenGL/RenderOpenGL.cpp

```diff
@@ -2711,205 +2814,586 @@ void RenderOpenGL::RenderTerrainD3D() {
+                    else
+                        uClipFlag = far_clip != 0 ? 5 : 0;
+                    static_sub_0048034E_stru_154.ClassifyPolygon(norm, Light_tile_dist);
+
+                    if (decal_builder->uNumDecals > 0)
+                        decal_builder->ApplyDecals(31 - pTilePolygon->dimming_level,
+                            4, &static_sub_0048034E_stru_154,
+                            3, VertexRenderList, 0,
+                            *(float*)&uClipFlag, -1);
+                    if (Lights.uNumLightsApplied > 0)
+                        lightmap_builder->ApplyLights(
+                            &Lights, &static_sub_0048034E_stru_154, 3,
+                            VertexRenderList, 0, uClipFlag);
+                }
 
 
-            if (norm_idx < 0 || norm_idx > uNumTerrainNormals - 1)
-                norm = 0;
-            else
-                norm = &pTerrainNormals[norm_idx];
 
-            // if (false) {
-            //    // if (engine->config->allow_lightmaps) {
-            //    // Ritor1: significant fps slowdown
-           //     lightmap_builder->StackLights_TerrainFace(
-            //       norm, &Light_tile_dist, VertexRenderList, 4,
-            //        1);
-            //    decal_builder->ApplyBloodSplatToTerrain(pTilePolygon, norm,
-            //        &Light_tile_dist, VertexRenderList, 4, 1);
-            //}
-
-
-            poly->uEdgeList1Size = x;
-            poly->uEdgeList2Size = z;
-
-            // draw animated water under shore
-            bool water_border_tile = false;
-            if (tile->IsWaterBorderTile()) {
-                glDepthMask(GL_FALSE);
-                {
-                    poly->texture =
-                        this->hd_water_tile_anim[this->hd_water_current_frame];
-
-                    auto texturew = (TextureOpenGL *)poly->texture;
-                    GLint thistexw = texturew->GetOpenGlTexture();
-
-                    // avoid rebinding same tex if we can
-                    if (lasttex != thistexw) {
-                        glBindTexture(GL_TEXTURE_2D, thistexw);
-                        lasttex = thistexw;
+
+
+
+
+                // pODMRenderParams->shading_dist_mist = temp;
+
+                // check the transparency and texture (tiles) mapping (проверка
+                // прозрачности и наложение текстур (тайлов))----------------------
+
+                if (tile->IsWaterTile()) {
+                    tile_texture = this->hd_water_tile_anim[this->hd_water_current_frame];
+                } else {
+                    tile_texture = tile->GetTexture();
+                }
+                // draw animated water under shore
+                bool water_border_tile = false;
+                if (tile->IsWaterBorderTile()) {
+                    glDepthMask(GL_FALSE);
+                    {
+                        pTilePolygon->texture =
+                            this->hd_water_tile_anim[this->hd_water_current_frame];
```

Comment:
btw side note again - if you check original game software render mode water (where it washes the shore) and compare it to original game accelerated mode, and then what we have - i think it's 3 completely different results (software mode afair was looking the best)

Reply:
Yeah i see what you mean, looks a little jerky.